pkg/services/anonymous/anonimpl/impl.go

```diff
@@ -2,178 +2,114 @@ package anonimpl
+		anonStore:  anonStore,
+		serverLock: serverLockService,
 	}
 
 	usageStats.RegisterMetricsFunc(a.usageStatFn)
 
-	return a
-}
-
-func (a *AnonDeviceService) usageStatFn(ctx context.Context) (map[string]any, error) {
-	anonDeviceCount, err := a.remoteCache.Count(ctx, string(anonymous.AnonDevice))
-	if err != nil {
-		return nil, nil
+	anonClient := &Anonymous{
+		cfg:               cfg,
+		log:               log.New("authn.anonymous"),
+		orgService:        orgService,
+		anonDeviceService: a,
 	}
 
-	authedDeviceCount, err := a.remoteCache.Count(ctx, string(anonymous.AuthedDevice))
-	if err != nil {
-		return nil, nil
+	if anonClient.cfg.AnonymousEnabled {
+		authBroker.RegisterClient(anonClient)
+		authBroker.RegisterPostLoginHook(a.untagDevice, 100)
 	}
 
-	anonUIDeviceCount, err := a.remoteCache.Count(ctx, string(anonymous.AnonDeviceUI))
-	if err != nil {
-		return nil, nil
-	}
+	return a
+}
 
-	authedUIDeviceCount, err := a.remoteCache.Count(ctx, string(anonymous.AuthedDeviceUI))
+func (a *AnonDeviceService) usageStatFn(ctx context.Context) (map[string]any, error) {
+	// Count the number of unique devices that have been updated in the last 30 days.
+	// One minute is added to the end time as mysql has a precision of seconds and it will break tests that write too fast.
+	anonUIDeviceCount, err := a.anonStore.CountDevices(ctx, time.Now().Add(-thirtyDays), time.Now().Add(time.Minute))
 	if err != nil {
-		return nil, nil
+		return nil, err
 	}
 
 	return map[string]any{
-		"stats.anonymous.session.count":   anonDeviceCount, // keep session for legacy data
-		"stats.users.device.count":        authedDeviceCount,
 		"stats.anonymous.device.ui.count": anonUIDeviceCount,
-		"stats.users.device.ui.count":     authedUIDeviceCount,
 	}, nil
 }
 
-func (a *AnonDeviceService) untagDevice(ctx context.Context, device *Device) error {
-	key, err := device.Key()
-	if err != nil {
-		return err
-	}
+func (a *AnonDeviceService) tagDeviceUI(ctx context.Context, httpReq *http.Request, device *anonstore.Device) error {
+	key := device.CacheKey()
 
-	if err := a.remoteCache.Delete(ctx, key); err != nil {
-		return err
+	if _, ok := a.localCache.Get(key); ok {
+		return nil
 	}
 
-	return nil
-}
+	a.localCache.SetDefault(key, struct{}{})
 
-func (a *AnonDeviceService) untagUIDevice(ctx context.Context, deviceID string, device *Device) error {
-	key, err := device.UIKey(deviceID)
-	if err != nil {
-		return err
+	if setting.Env == setting.Dev {
+		a.log.Debug("Tagging device for UI", "deviceID", device.DeviceID, "device", device, "key", key)
 	}
 
-	if err := a.remoteCache.Delete(ctx, key); err != nil {
+	if err := a.anonStore.CreateOrUpdateDevice(ctx, device); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (a *AnonDeviceService) tagDeviceUI(ctx context.Context, httpReq *http.Request, device Device) error {
-	deviceID := httpReq.Header.Get(deviceIDHeader)
+func (a *AnonDeviceService) untagDevice(ctx context.Context,
```

Comment:
We should not untag a device if there was an error during the login proccess. Then a failed login would untag the device